### PR TITLE
Get var consequences return list 832

### PR DESCRIPTION
**Pull request by rcurrie**
Here's a pull request for the whole kit and kaboodle:
- Dockerize splicing pipeline
- Move all computation internal (no calls to ensembl or ucsc browser)
- Switch to list for var consequences
- Changes Marc made (thanks!) to make CI happy
This matches the 0.0.3 image pushed to docker hub which created the output used for concordance.

### CODE REVIEW

File: pipeline/splicing/calcVarPriors.py [hunk 1]
```diff
@@ -722,23 +703,15 @@ def getFastaSeq(chrom, rangeStart, rangeStop, plusStrandSeq=True):
...
-    
-    lines = req.content.split('\n')
-    # because sequence is located at index 5 in dictionary
-    sequence = lines[5]
-    for base in sequence:
-        assert base in ["A", "C", "G", "T", "a", "c", "g", "t"]
-    if plusStrandSeq == True:
-        return sequence.upper()
+    # NOTE: pyfaidx is NOT thread safe. Would be better to have one
+    # REMIND: Switch to one per thread/process
```
Comment (zfisch) [inline pipeline/splicing/calcVarPriors.py]: sounds precarious -- is this something that should be fixed before v1?

Comment (zfisch) [reply pipeline/splicing/calcVarPriors.py]: also, if we use it:

> If you use pyfaidx in your publication, please cite:
> 
> Shirley MD, Ma Z, Pedersen B, Wheelan S. Efficient "pythonic" access to FASTA files using pyfaidx. PeerJ PrePrints 3:e1196. 2015.

File: pipeline/splicing/calcVarPriors.py [hunk 2]
```diff
@@ -1798,7 +1774,10 @@ def getPriorProbAfterGreyZoneSNS(variant, boundaries):
                 "inExonicPortionFlag": "N/A",
                 "CIDomainInRegionFlag": "N/A",
                 "isDivisibleFlag": "N/A",
-                "lowMESFlag": "N/A"}
+                "lowMESFlag": "N/A",
+                "varConsequences": ",".join(varCons)}
+
+    assert False, "Should never reach this"
```
Comment (zfisch) [inline pipeline/splicing/calcVarPriors.py]: Why? Any ideas on more information to provide in the case that it does reach this?

File: pipeline/splicing/calcVarPriors.py [hunk 3]
```diff
@@ -3444,65 +3428,112 @@ def getVarDict(variant, boundaries):
...
-                        help="Specifies which boundaries ('enigma' or 'priors') to use for clinically important domains")
-    parser.add_argument('-g', "--genomeFile", help="Fasta file containing hg38 reference genome")
-    parser.add_argument('-t', "--transcriptFile", help="RefSeq annotation hg38-based genepred file")
-    args = parser.parse_args()    
-
-    inputData = csv.DictReader(open(args.inputFile, "r"), delimiter="\t")
+
+# Globals accessed in multi-processing function calc_one
+# REMIND: Use functools.partial to pass direction into the map function
+genome38 = None  # This is actually never used...
```
Comment (zfisch) [inline pipeline/splicing/calcVarPriors.py]: can we remove it altogether?

File: pipeline/splicing/calcVarPriors.py [hunk 4]
```diff
@@ -3444,65 +3428,112 @@ def getVarDict(variant, boundaries):
...
-    parser.add_argument('-b', "--boundaries", default="enigma",
-                        help="Specifies which boundaries ('enigma' or 'priors') to use for clinically important domains")
-    parser.add_argument('-g', "--genomeFile", help="Fasta file containing hg38 reference genome")
-    parser.add_argument('-t', "--transcriptFile", help="RefSeq annotation hg38-based genepred file")
-    args = parser.parse_args()    
-
-    inputData = csv.DictReader(open(args.inputFile, "r"), delimiter="\t")
+
+# Globals accessed in multi-processing function calc_one
+# REMIND: Use functools.partial to pass direction into the map function
```
Comment (zfisch) [inline pipeline/splicing/calcVarPriors.py]: now or later?

File: pipeline/splicing/calcVarPriors.py [hunk 5]
```diff
@@ -3444,65 +3428,112 @@ def getVarDict(variant, boundaries):
...
-                  "refRefAccZ", "altRefAccMES", "altRefAccZ", "refRefAccSeq", "altRefAccSeq", "refAccVarStart", "refAccVarLength", "refAccExonStart",
-                  "refAccIntronStart", "refDeNovoAccMES", "refDeNovoAccZ", "altDeNovoAccMES", "altDeNovoAccZ", "refDeNovoAccSeq", "altDeNovoAccSeq",
-                  "deNovoAccVarStart", "deNovoAccVarLength", "deNovoAccExonStart", "deNovoAccIntronStart", "deNovoAccGenomicSplicePos",
-                  "deNovoAccTranscriptSplicePos", "closestAccGenomicSplicePos", "closestAccTranscriptSplicePos", "closestAccRefMES",
-                  "closestAccRefZ", "closestAccRefSeq", "closestAccAltMES", "closestAccAltZ", "closestAccAltSeq",
-                  "closestAccExonStart", "closestAccIntronStart", "deNovoAccAltGreaterRefFlag", "deNovoAccAltGreaterClosestRefFlag",
-                  "deNovoAccAltGreaterClosestAltFlag", "deNovoAccFrameshiftFlag", "spliceSite", "spliceRescue", "spliceFlag", "frameshiftFlag",
-                  "inExonicPortionFlag", "CIDomainInRegionFlag", "isDivisibleFlag", "lowMESFlag"]
+    newHeaders = open("headers.tsv", "r").read().split()
```
Comment (zfisch) [inline pipeline/splicing/calcVarPriors.py]: 👍

File: pipeline/splicing/calcVarPriors.py [hunk 6]
```diff
@@ -722,23 +703,15 @@ def getFastaSeq(chrom, rangeStart, rangeStop, plusStrandSeq=True):
         regionStart = rangeStop
         regionEnd = rangeStart
     
-    url = "http://genome.ucsc.edu/cgi-bin/das/hg38/dna?segment=%s:%d,%d" % (chrom, regionStart, regionEnd)
```
Comment (zfisch) [inline pipeline/splicing/calcVarPriors.py]: 👍

File: pipeline/splicing/calcVarPriors.py [hunk 7]
```diff
@@ -3444,65 +3428,112 @@ def getVarDict(variant, boundaries):
...
-    with open(args.transcriptFile) as infile:
-        transcripts = pyhgvs_utils.read_transcripts(infile)
+    transcripts = pyhgvs_utils.read_transcripts(transcripts)
+
+    brca1Transcript = transcripts.get(BRCA1_RefSeq)
+    brca2Transcript = transcripts.get(BRCA2_RefSeq)
+
+    # Create a pool of processes and calculate in parallel
+    click.echo("Processing using {} processes".format(processes), err=True)
+    pool = multiprocessing.Pool(processes)
```
Comment (zfisch) [inline pipeline/splicing/calcVarPriors.py]: 👍

File: pipeline/splicing/calcVarPriors.py [hunk 8]
```diff
@@ -3444,65 +3428,112 @@ def getVarDict(variant, boundaries):
...
+    try:
+        # Normal map has a bug if there is no timout that prevents Keyboard interrupts:
+        # https://stackoverflow.com/questions/1408356/keyboard-interrupts-with-pythons-multiprocessing-pool/1408476#1408476
+        calculatedVariants = pool.map_async(calc_one, list(inputData)).get(99999999)
```
Comment (zfisch) [inline pipeline/splicing/calcVarPriors.py]: 🤷‍♂️

File: pipeline/splicing/calcVarPriors.py [hunk 9]
```diff
@@ -7,19 +7,22 @@
...
 import csv
 import requests
-import sys
 import time
 import json
-import re
 import subprocess
-import tempfile
-import os
+import click
```
Comment (zfisch) [inline pipeline/splicing/calcVarPriors.py]: cool!

File: pipeline/splicing/references.sh [hunk 10]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
+# Download and verify references - to be called from within the docker
+echo "Downloading references. This may take 1-2 hours..."
+wget -N -P /references http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
```
Comment (zfisch) [inline pipeline/splicing/references.sh]: we also host this file at http://brcaexchange.org/backend/downloads/resources/hg38.fa.gz, not sure if it matters where we get it

File: pipeline/splicing/test_calcVarPriors.py [hunk 11]
```diff
@@ -242,39 +242,39 @@ def test_getVarType(self, checkSequence):
...
-        self.assertEquals(varCons, "unable_to_determine")
-
-        self.variant["Alt"] = "-"
-        varCons = calcVarPriors.getVarConsequences(self.variant)
-        self.assertEquals(varCons, "unable_to_determine")
-
-        self.variant["Alt"] = "38413620"
-        varCons = calcVarPriors.getVarConsequences(self.variant)
-        self.assertEquals(varCons, "unable_to_determine")
+    # def test_getVarConsequences(self):
```
Comment (zfisch) [inline pipeline/splicing/test_calcVarPriors.py]: what's the status of these tests? should they remain quoted, be deleted, or be reinstated?

File: pipeline/splicing/test_calcVarPriors.py [hunk 12]
```diff
@@ -7416,7 +7416,7 @@ def test_getPriorProbInExonSNSEnigmaCIDeNovoAcceptorBRCA2(self, getVarLocation,
     @mock.patch('calcVarPriors.getVarType', return_value = varTypes["sub"])
     @mock.patch('calcVarPriors.getPriorProbProteinSNS', return_value = {'enigmaClass': 'class_5',
                                                                         'priorProb': 0.99})
-    @mock.patch('calcVarPriors.getVarConsequences', return_value = "stop_gained")
+    @mock.patch('calcVarPriors.getVarConsequences', return_value = ["stop_gained"])
```
Comment (zfisch) [inline pipeline/splicing/test_calcVarPriors.py]: looks like there's a bunch of trailing spaces that can be removed in this file

**Comment by zfisch** [comment]
Added @falquaddoomi to code review since he's volunteered to work on structural variant priors.

**Comment by rcurrie** [comment]
@zfisch I'm definitely interested in resolving the comments an appreciate you suffering through my changes :-) I'll work through them and comment. With respect to the big files I'll pull them. We have the 88 samples for testing that should do the trick.

File: .circleci/config.yml [hunk 13]
```diff
@@ -93,23 +79,122 @@ jobs:
...
+           name: Load Docker image layer cache
+           command: |
+             set +o pipefail
+             docker load -i /caches/pipeline.tar | true
+       - run:
+           name: Build application Docker image
+           working_directory: /pipeline/pipeline/docker
+           command: |
+             pwd
+             echo $(pwd)
```
[the remark this answers is not in the document]
Comment (rcurrie) [reply .circleci/config.yml]: @izcram Ok to remove?

File: .circleci/config.yml [hunk 14]
```diff
@@ -93,23 +79,122 @@ jobs:
...
              tags:
                only: /.*/
+       - test-splicing-pipeline:
+          filters:
+             tags:
+               only: /.*/
        - deploy-dev:
            requires:
-             - build
+             - buildweb
```
[the remark this answers is not in the document]
Comment (rcurrie) [reply .circleci/config.yml]: @izcram ?

File: pipeline/splicing/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    python-pip \
+    python-dev \
+    libmysqlclient-dev
+
+WORKDIR /app
+
+# pyhgvs 0.9.4
+RUN pip install git+git://github.com/counsyl/hgvs.git@aebe5bd9683f4b5937fd653ce4b13fcd4f3ebb10
+
+# RUN pip install --upgrade pip NOTE: pyhgvs breaks with pip > v8
```
[the remark this answers is not in the document]
Comment (rcurrie) [reply pipeline/splicing/Dockerfile]: Yup, still relevant as upgrading pip bumps up to something 18 and breaks. We're using an older pyhgvs here and I was reticent to not change until we made it through this initial validation. I think there has been discussion about upgrading? I'm inclined to wait until we integrate and put this into production and then do the upgrade along with the next step (structural...)

File: pipeline/splicing/calcVarPriors.py (hunk 3 above)
Comment (rcurrie) [reply pipeline/splicing/calcVarPriors.py]: Went ahead and removed this global and the initial calling. Below that genome is passed through a ton of functions (all None and working) and I'd prefer to leave it for the bit refactor after this.

File: pipeline/splicing/calcVarPriors.py (hunk 4 above)
Comment (rcurrie) [reply pipeline/splicing/calcVarPriors.py]: Done!

File: pipeline/splicing/test_calcVarPriors.py (hunk 11 above)
Comment (rcurrie) [reply pipeline/splicing/test_calcVarPriors.py]: Deleted: They were for the older version that just returned a single value

File: .circleci/config.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (izcram) [reply .circleci/config.yml]: yes sure, sorry for the dangling debugging 'code'

File: .circleci/config.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (izcram) [reply .circleci/config.yml]: The tests are run whenever CI gets triggered, e.g. in a  PR or tagging. see also https://github.com/BRCAChallenge/brca-exchange/blob/1337c86d5958cf98e36f0cc4166cb23d3ef6f8f5/.circleci/config.yml#L180

File: pipeline/splicing/Makefile [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+debug:
+	# Run the docker mapping out of local directory for development
+	docker run --rm -it \
+		--entrypoint /bin/bash \
+		--user=`id -u`:`id -g` \
+		-v `readlink -f $(REFERENCES)`:/references \
+		-v `pwd`:/app \
+	  $(IMAGE_NAME)
+
```
Comment (izcram) [inline pipeline/splicing/Makefile]: seems the `unit_tests_docker` target was not so dead after all ;) If you add it back, circleci should run through. Could you also add a comment, that it is used for circleci, which I should have done anyways?

File: .circleci/config.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (rcurrie) [reply .circleci/config.yml]: Done.

**Comment by zfisch** [comment]
Hey @rcurrie @izcram do you think we could get this PR resolved and merged by tomorrow? That will give us some time to integrate it fully into the automated pipeline and get some new data for @falquaddoomi to work with for the backend/UI. We're hoping to get something up to beta by the end of the week, but if we need a little more time I'm sure we can push it back a little.

Please let me know if there's anything I can do to help us get there!

**Comment by rcurrie** [comment]
Yes! I’d like to get it put to bed today. Making a few more changes to address comments. At your service to discuss or resolve anything additional.

File: pipeline/splicing/Makefile (hunk 16 above)
Comment (rcurrie) [reply pipeline/splicing/Makefile]: Doh! I've put it back and tested plus comment

File: .circleci/config.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (rcurrie) [reply .circleci/config.yml]: I've removed the 2 references large files:

built_with_mupit.tsv  (25.4 MB) 
built_with_priors.tsv  (33.4 MB)

Using [BFG](https://rtyley.github.io/bfg-repo-cleaner/) (first time I've used it - pretty nice). For reference the top large files in the overall repo are:

```
    Filename                      Git id
        --------------------------------------------------------------------
        1000G_brca.hg38.vcf         | 6b5b6e2e (55.9 MB)
        1000G_brca.hg38_toy.vcf     | b31b01de (11.1 MB)
        RefSeq.genepred             | e5327609 (49.3 MB)
        built_with_change_types.tsv | 34095970 (14.2 MB), a2b56343 (15.0 MB)
        built_with_mupit.tsv        | dc4fc70c (25.4 MB)
        built_with_priors.tsv       | 09b2cec5 (33.4 MB)
        release-10-21-16.tar.gz     | f9ecb0c7 (26.1 MB), e11ea1ca (26.1 MB)
        release-11-05-16.tar.gz     | faf7d79d (27.0 MB)
```

**Pull request merged**